In [22]:
import re

<link href="style.css" rel="stylesheet"></link>

<article class="day-desc read-aloud"><h2>--- Day 9: Mirage Maintenance ---</h2><p>You ride the camel through the sandstorm and stop where the ghost's maps told you to stop. <span title="The sound of a sandstorm slowly settling.">The sandstorm subsequently subsides, somehow seeing you standing at an <em>oasis</em>!</span></p>
<p>The camel goes to get some water and you stretch your neck. As you look up, you discover what must be yet another giant floating island, this one made of metal! That must be where the <em>parts to fix the sand machines</em> come from.</p>
<p>There's even a <a href="https://en.wikipedia.org/wiki/Hang_gliding" target="_blank">hang glider</a> partially buried in the sand here; once the sun rises and heats up the sand, you might be able to use the glider and the hot air to get all the way up to the metal island!</p>
<p>While you wait for the sun to rise, you admire the oasis hidden here in the middle of Desert Island. It must have a delicate ecosystem; you might as well take some ecological readings while you wait. Maybe you can report any environmental instabilities you find to someone so the oasis can be around for the next sandstorm-worn traveler.</p>
<p>You pull out your handy <em>Oasis And Sand Instability Sensor</em> and analyze your surroundings. The OASIS produces a report of many values and how they are changing over time (your puzzle input). Each line in the report contains the <em>history</em> of a single value. For example:</p>
<pre><code>0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
</code></pre>
<p>To best protect the oasis, your environmental report should include a <em>prediction of the next value</em> in each history. To do this, start by making a new sequence from the <em>difference at each step</em> of your history. If that sequence is <em>not</em> all zeroes, repeat this process, using the sequence you just generated as the input sequence. Once all of the values in your latest sequence are zeroes, you can extrapolate what the next value of the original history should be.</p>
<p>In the above dataset, the first history is <code>0 3 6 9 12 15</code>. Because the values increase by <code>3</code> each step, the first sequence of differences that you generate will be <code>3 3 3 3 3</code>. Note that this sequence has one fewer value than the input sequence because at each step it considers two numbers from the input. Since these values aren't <em>all zero</em>, repeat the process: the values differ by <code>0</code> at each step, so the next sequence is <code>0 0 0 0</code>. This means you have enough information to extrapolate the history! Visually, these sequences can be arranged like this:</p>
<pre><code>0   3   6   9  12  15
  3   3   3   3   3
    0   0   0   0
</code></pre>
<p>To extrapolate, start by adding a new zero to the end of your list of zeroes; because the zeroes represent differences between the two values above them, this also means there is now a placeholder in every sequence above it:</p><p>
</p><pre><code>0   3   6   9  12  15   <em>B</em>
  3   3   3   3   3   <em>A</em>
    0   0   0   0   <em>0</em>
</code></pre>
<p>You can then start filling in placeholders from the bottom up. <code>A</code> needs to be the result of increasing <code>3</code> (the value to its left) by <code>0</code> (the value below it); this means <code>A</code> must be <code><em>3</em></code>:</p>
<pre><code>0   3   6   9  12  15   B
  3   3   3   3   <em>3</em>   <em>3</em>
    0   0   0   0   <em>0</em>
</code></pre>
<p>Finally, you can fill in <code>B</code>, which needs to be the result of increasing <code>15</code> (the value to its left) by <code>3</code> (the value below it), or <code><em>18</em></code>:</p>
<pre><code>0   3   6   9  12  <em>15</em>  <em>18</em>
  3   3   3   3   3   <em>3</em>
    0   0   0   0   0
</code></pre>
<p>So, the next value of the first history is <code><em>18</em></code>.</p>
<p>Finding all-zero differences for the second history requires an additional sequence:</p>
<pre><code>1   3   6  10  15  21
  2   3   4   5   6
    1   1   1   1
      0   0   0
</code></pre>
<p>Then, following the same process as before, work out the next value in each sequence from the bottom up:</p>
<pre><code>1   3   6  10  15  21  <em>28</em>
  2   3   4   5   6   <em>7</em>
    1   1   1   1   <em>1</em>
      0   0   0   <em>0</em>
</code></pre>
<p>So, the next value of the second history is <code><em>28</em></code>.</p>
<p>The third history requires even more sequences, but its next value can be found the same way:</p>
<pre><code>10  13  16  21  30  45  <em>68</em>
   3   3   5   9  15  <em>23</em>
     0   2   4   6   <em>8</em>
       2   2   2   <em>2</em>
         0   0   <em>0</em>
</code></pre>
<p>So, the next value of the third history is <code><em>68</em></code>.</p>
<p>If you find the next value for each history in this example and add them together, you get <code><em>114</em></code>.</p>
<p>Analyze your OASIS report and extrapolate the next value for each history. <em>What is the sum of these extrapolated values?</em></p>
</article>


In [23]:
def extrapolate(history):
    h = [history]
    while any(v != 0 for v in h[-1]):
        h.append([h[-1][i] - h[-1][i - 1] for i in range(1, len(h[-1]))])

    for i in range(len(h) - 2, -1, -1):
        h[i - 1].append(h[i][-1] + h[i - 1][-1])
    return h[0][-1]


extrapolate([0, 3, 6, 9, 12, 15]) == 18
extrapolate([1, 3, 6, 10, 15, 21]) == 28
extrapolate([10, 13, 16, 21, 30, 45]) == 68

True

In [24]:
histories = """
0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
"""


def extrapolate_histories(histories: str) -> int:
    hists = [[int(i) for i in line.split()] for line in histories.strip().splitlines()]
    return sum(extrapolate(h) for h in hists)


assert extrapolate_histories(histories) == 114

In [25]:
with open("./input/day9.txt") as f:
    print(extrapolate_histories(f.read()))

1930746032


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>1930746032</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Of course, it would be nice to have <em>even more history</em> included in your report. Surely it's safe to just <em>extrapolate backwards</em> as well, right?</p>
<p>For each history, repeat the process of finding differences until the sequence of differences is entirely zero. Then, rather than adding a zero to the end and filling in the next values of each previous sequence, you should instead add a zero to the <em>beginning</em> of your sequence of zeroes, then fill in new <em>first</em> values for each previous sequence.</p>
<p>In particular, here is what the third example history looks like when extrapolating back in time:</p>
<pre><code><em>5</em>  10  13  16  21  30  45
  <em>5</em>   3   3   5   9  15
   <em>-2</em>   0   2   4   6
      <em>2</em>   2   2   2
        <em>0</em>   0   0
</code></pre>
<p>Adding the new values on the left side of each sequence from bottom to top eventually reveals the new left-most history value: <code><em>5</em></code>.</p>
<p>Doing this for the remaining example data above results in previous values of <code><em>-3</em></code> for the first history and <code><em>0</em></code> for the second history. Adding all three new values together produces <code><em>2</em></code>.</p>
<p>Analyze your OASIS report again, this time extrapolating the <em>previous</em> value for each history. <em>What is the sum of these extrapolated values?</em></p>
</article>


In [26]:
from tabulate import tabulate


def pre_extrapolate(history):
    h = [history]
    while any(v != 0 for v in h[-1]):
        h.append([h[-1][i] - h[-1][i - 1] for i in range(1, len(h[-1]))])

    for i in range(len(h) - 2, -1, -1):
        h[i].insert(0, h[i][0] - h[i + 1][0])

    return h[0][0]


assert pre_extrapolate([0, 3, 6, 9, 12, 15]) == -3
assert pre_extrapolate([1, 3, 6, 10, 15, 21]) == 0
assert pre_extrapolate([10, 13, 16, 21, 30, 45]) == 5

In [27]:
histories = """
0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
"""


def pre_extrapolate_histories(histories: str) -> int:
    hists = [[int(i) for i in line.split()] for line in histories.strip().splitlines()]
    return sum(pre_extrapolate(h) for h in hists)


assert pre_extrapolate_histories(histories) == 2

In [28]:
with open("./input/day9.txt") as f:
    print(pre_extrapolate_histories(f.read()))

1154


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>1154</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2023">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="9/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Mirage+Maintenance%22+%2D+Day+9+%2D+Advent+of+Code+2023&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F9&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Mirage+Maintenance%22+%2D+Day+9+%2D+Advent+of+Code+2023+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F9';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
